shell/server/core.js

```diff
@@ -112,33 +116,35 @@ NotificationHandle.prototype.close = function () {
   });
 };
 
-NotificationHandle.prototype.save = function (params) {
-  var self = this;
+saveFrontendRef = function (frontendRef, owner) {
   return inMeteor(function () {
-    var sturdyRef = new Buffer(Random.id(20));
+    var sturdyRef = new Buffer(generateSturdyRef());
     var hashedSturdyRef = hashSturdyRef(sturdyRef);
     ApiTokens.insert({
       _id: hashedSturdyRef,
-      frontendRef: {
-        notificationHandle: self.notificationId
-      },
-      owner: params.sealFor,
+      frontendRef: frontendRef,
+      owner: owner,
+      created: new Date()
     });
-    self.saved = true;
     return {sturdyRef: sturdyRef};
   });
 };
 
-function restoreInternal(sturdyRef, ownerPattern) {
-  // Restores `sturdyRef`, checking first that its owner matches `ownerPattern`.
+NotificationHandle.prototype.save = function (params) {
+  return saveFrontendRef({notificationHandle: this.notificationId}, params.sealFor);
+};
 
+restoreInternal = function (sturdyRef, ownerPattern, requiredPermissions) {
```

Comment:
I think you can get away without the `requiredPermissions` parameter here. To check the permissions for a `grain.introducerUser` owner, you can use [`Match.Where`](http://docs.meteor.com/#/full/matchpatterns). The other permissions you'll need to check will be in the `token.requirements` list; they won't depend on any parameters to this function.